[WFLY-10711] Support Eclipse MicroProfile Health

jmesnil opened the pull request:
* add microprofile-smallrye-health extension
* the MicroProfile Health implementation is provided by smallrye-health
  1.0.2
* Healthiness can be queried by:
  * the `check` operation on the `/subsystem=microprofile-health-smallrye` resource
  * the HTTP endpoint http://localhost:9990/health

The HTTP endpoint is attached to WildFly HTTP management interface.
It secure access is controlled by the `security-enabled` attribute.
Default value i set to `true` (i.e. authentication is required) but the
attribute is set to `false` in the standalone configuration.

JIRA: https://issues.jboss.org/browse/WFLY-10711
Proposal: https://github.com/wildfly/wildfly-proposals/pull/105

This PR has a downstream dependency on WFCORE-3999 that has been submitted in https://github.com/wildfly/wildfly-core/pull/3427

----
Please make sure your PR meets the following requirements:
- [x] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue(s)
- [x] Pull Request contains description of the issue(s)
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [x] Pull Request requires a change to the documentation
- [x] Documentation have been updated accordingly
- [x] Tests were added to cover changes

For new features ensure as well:
- [x] Analysis was done
- [x] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

jmesnil committed:
[WFLY-10711] Support Eclipse MicroProfile Health

* add microprofile-smallrye-health extension
* the MicroProfile Health implementation is provided by smallrye-health
  1.0.2
* Healthiness can be queried by:
  * the `check` operation on the `/subsystem=microprofile-health-smallrye` resource
  * the HTTP endpoint http://localhost:9990/health

The HTTP endpoint is attached to WildFly HTTP management interface.
It secure access is controlled by the `security-enabled` attribute.
Default value i set to `true` (i.e. authentication is required) but the
attribute is set to `false` in the standalone configuration.

JIRA: https://issues.jboss.org/browse/WFLY-10711

jmesnil commented:
@rsvoboda FIY, this PR requires WildFly Core from https://github.com/jmesnil/wildfly-core/tree/WFCORE-3999_HttpHandler_ExtensibleHttpManagement

When everything is built, workflow is

1. unzip server
2. start server
3. http://localhost:9990/health

The HTTP endpoint is *not* secured in the standalone configuration.

rsvoboda commented:
Tracking my concerns here: https://issues.jboss.org/issues/?jql=project%20%3D%20WFWIP%20AND%20component%20%3D%20%22MP%20Health%22%20AND%20resolution%20is%20EMPTY%20

WFWIP-97 is the most important, WFWIP-98 needs some clarification

jmesnil committed:
[WFLY-10711] Support Eclipse MicroProfile Health

Remove services when the subsystem is removed.

JIRA: https://issues.jboss.org/browse/WFLY-10711
JIRA: https://issues.jboss.org/browse/WFWIP-99

jmesnil committed:
[WFLY-10711] Support Eclipse MicroProfile Health

* add microprofile-profile-health subsystem to the standalone HA
  profiles

JIRA: https://issues.jboss.org/browse/WFLY-10711
JIRA: https://issues.jboss.org/browse/WFWIP-97

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Health.adoc:
```diff
@@ -0,0 +1,86 @@
+{"outcome":"DOWN","checks":[{"name":"myFailingProbe","state":"DOWN","data":{"foo":"bar"}}]}
+----
+
+If security has been enabled, the HTTP client must pass the credentials corresponding to a manager user
```
s/manager/management

jmesnil replied:
fixed

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/io/smallrye/health/main/module.xml:
```diff
@@ -0,0 +1,37 @@
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.8" name="io.smallrye.health">
```
Is this one private?

jmesnil replied:
you're right, I flagged it as `private`

jmesnil replied:
yes, I amended the definition

bstansberry commented on microprofile/health-smallrye/src/main/resources/org/wildfly/extension/microprofile/health/LocalDescriptions.properties:
```diff
@@ -0,0 +1,5 @@
+microprofile-health-smallrye.add=Add the subsystem
+microprofile-health-smallrye.remove=Remove the subsystem
+microprofile-health-smallrye.check=Check the healthiness of the application server and its deployments
+microprofile-health-smallrye.security-enabled=True if authentication is required to access the HTTP endpoint.
```
Perhaps add "on the management interface."?

If we later expose this on the undertow subsystem having this wording here gives us some freedom to use some different configuration for authentication there if needed. And if it turns out this attribute is applicable we can just remove these words.

No big deal because we could always add those words later if it turned out this attribute could only apply to the mgmt interface.

jmesnil replied:
fixed

bstansberry commented:
I had one trivial substantive correction and a couple minor questions but overall this looks good. Except for needing a rebase. :)

rsvoboda commented:
@jmesnil all the checks are failing here, maybe you can squash those 3 commits into one, as a bonus TC jobs will be triggered again :)

jmesnil commented:
retest this please

jmesnil commented:
PR pull player is acting quite strange. All runs can be checked from https://ci.wildfly.org/project.html?projectId=WF_PullRequest&branch_WF_PullRequest=pull%2F11503

jmesnil commented:
All requirements are met (removing missing-reqs label)

jmesnil commented:
retest this please

jmesnil commented:
@wildfly/mergers-full PR has been rebased and tests pass. Failures in Linux JDK10 are not related to the PR.
